fuel-merkle/src/common/position.rs

```diff
@@ -110,24 +109,44 @@ impl Position {
 
     /// Construct a position from a leaf index. The in-order index corresponding
     /// to the leaf index will always equal the leaf index multiplied by 2.
-    pub fn from_leaf_index(index: u64) -> Self {
-        Position(index * 2)
+    /// Panics if index is too large to fit into u64.
+    pub fn from_leaf_index(index: u64) -> Option<Self> {
+        Some(Position(index.checked_mul(2)?))
+    }
+
+    #[cfg(test)]
+    pub(crate) fn from_leaf_index_unwrap(index: u64) -> Self {
+        Self::from_leaf_index(index).expect("Index too large")
     }
 
     /// The sibling position.
     /// A position shares the same parent and height as its sibling.
     pub fn sibling(self) -> Self {
-        let shift = 1 << (self.height() + 1);
-        let index = self.in_order_index() as i64 + shift * self.direction();
-        Self::from_in_order_index(index as u64)
+        #[allow(clippy::arithmetic_side_effects)]
+        let shift = 1 << (self.height() + 1); // height() < u32::MAX
+        let this = self.in_order_index();
+        #[allow(clippy::arithmetic_side_effects)]
+        Self::from_in_order_index(match self.orientation() {
+            // height in relation to in_order_index cannot underflow
+            Side::Left => this - shift,
+            // it's not possible to actually construct a tree so large that this would
+            // overflow
+            Side::Right => this + shift,
+        })
     }
 
     /// The parent position.
     /// The parent position has a height less 1 relative to this position.
     pub fn parent(self) -> Self {
         let shift = 1 << self.height();
-        let index = self.in_order_index() as i64 + shift * self.direction();
-        Self::from_in_order_index(index as u64)
+        let this = self.in_order_index();
+        #[allow(clippy::arithmetic_side_effects)]
+        Self::from_in_order_index(match self.orientation() {
+            // height in relation to in_order_index cannot underflow
+            Side::Left => this - shift,
+            // it's not possible to actually construct a tree so larget his would overflow
```

Comment:
I don't think that we can guarantee that this function will not overflow since it is public, and the `Postiion` could be created with an incorrect value for `from_in_order_index`.

Reply:
Ah right. Fixed in ced9f8cd.